src/ucm/malloc/malloc_hook.c

```diff
@@ -124,46 +120,64 @@ static ucm_malloc_hook_state_t ucm_malloc_hook_state = {
     .free             = NULL,
     .heap_start       = (void*)-1,
     .heap_end         = (void*)-1,
-    .ptrs             = KHASH_STATIC_INITIALIZER,
+    .mmap_pages       = KHASH_STATIC_INITIALIZER,
     .env_lock         = PTHREAD_MUTEX_INITIALIZER,
     .env_strs         = NULL,
     .num_env_strs     = 0
 };
 
 int ucm_dlmallopt_get(int); /* implemented in ptmalloc */
 
+static int64_t ucm_malloc_page_address(void *ptr)
+{
+    return ucs_align_down_pow2((uintptr_t)ptr, ucm_get_page_size());
+}
+
 static void ucm_malloc_mmaped_ptr_add(void *ptr)
 {
-    int hash_extra_status;
-    khiter_t hash_it;
+    khiter_t khiter;
+    int khret;
 
     ucs_recursive_spin_lock(&ucm_malloc_hook_state.lock);
 
-    hash_it = kh_put(mmap_ptrs, &ucm_malloc_hook_state.ptrs, ptr,
-                     &hash_extra_status);
-    ucm_assert_always(hash_extra_status >= 0);
-    ucm_assert_always(hash_it != kh_end(&ucm_malloc_hook_state.ptrs));
+    khiter = kh_put(mmap_pages, &ucm_malloc_hook_state.mmap_pages,
+                    ucm_malloc_page_address(ptr), &khret);
+    if (khret == UCS_KH_PUT_KEY_PRESENT) {
+        ++kh_value(&ucm_malloc_hook_state.mmap_pages, khiter);
+    } else if ((khret == UCS_KH_PUT_BUCKET_EMPTY) ||
+               (khret == UCS_KH_PUT_BUCKET_CLEAR)) {
+        kh_value(&ucm_malloc_hook_state.mmap_pages, khiter) = 1;
+    } else {
+        ucm_fatal("kh_put failed: %d", khret);
+    }
 
     ucs_recursive_spin_unlock(&ucm_malloc_hook_state.lock);
 }
 
 static int ucm_malloc_mmaped_ptr_remove_if_exists(void *ptr)
 {
-    khiter_t hash_it;
-    int found;
+    khiter_t khiter;
+    size_t *count;
 
     ucs_recursive_spin_lock(&ucm_malloc_hook_state.lock);
 
-    hash_it = kh_get(mmap_ptrs, &ucm_malloc_hook_state.ptrs, ptr);
-    if (hash_it == kh_end(&ucm_malloc_hook_state.ptrs)) {
-        found = 0;
-    } else {
-        found = 1;
-        kh_del(mmap_ptrs, &ucm_malloc_hook_state.ptrs, hash_it);
+    khiter = kh_get(mmap_pages, &ucm_malloc_hook_state.mmap_pages,
+                    ucm_malloc_page_address(ptr));
+    if (khiter == kh_end(&ucm_malloc_hook_state.mmap_pages)) {
+        ucs_recursive_spin_unlock(&ucm_malloc_hook_state.lock);
+        return 0;
+    }
+
+    /* Decrement the hash value that counts page allocations */
+    count = &kh_value(&ucm_malloc_hook_state.mmap_pages, khiter);
+    if (--*count <= 0) {
+        /* Remove hash entry if all its allocations were released */
+        ucm_assert(*count == 0);
+        kh_del(mmap_pages, &ucm_malloc_hook_state.mmap_pages, khiter);
```

Comment:
maybe?
```suggestion
    count = &kh_value(&ucm_malloc_hook_state.mmap_pages, khiter);
    ucm_assert(*count > 0);
    if (--*count == 0) {
        /* Remove hash entry if all its allocations were released */
        kh_del(mmap_pages, &ucm_malloc_hook_state.mmap_pages, khiter);
```

or it was intent to compare `*count` with `0` in `ucm_assert`?
but `count` is always >= 0, because it `size_t`, i.e. unsigned

Reply:
fixed (force push due to small PR)